### PR TITLE
Direct3D 12 bug-fixes

### PR DESCRIPTION
**Pull request by slomp**
All known issues have been fixed.
It's ready to merge after the review.
We'll need to re-enable d3d12 in the automated tests too.

### CODE REVIEW

File: src/runtime/hashmap.h [hunk 1]
```diff
@@ -6,6 +6,14 @@
...
 #include "scoped_mutex_lock.h"
 
+#ifndef hashmap_malloc
+#define hashmap_malloc(size)    halide_malloc(NULL, size)
+#endif//hashmap_malloc
+
+#ifndef hashmap_free
+#define hashmap_free(memory)    halide_free(NULL, memory)
+#endif//hashmap_free
+
```
Comment (shoaibkamil) [inline src/runtime/hashmap.h]: I'm not sure I understand why this change; is everything the same as before, except now there's an extra preprocessor definition?

File: src/runtime/d3d12compute.cpp [hunk 2]
```diff
@@ -194,6 +197,10 @@ static T zero_struct() {
     return zero;
 }
 
+#define hashmap_malloc  d3d12_malloc
+#define hashmap_free    d3d12_free
+#include "hashmap.h"
+
```
Comment (shoaibkamil) [inline src/runtime/d3d12compute.cpp]: Ah, now I understand why there's a preprocessor definition in `hashmap.h`.  Please add an explanatory comment there so nobody accidentally changes it.

File: src/runtime/d3d12compute.cpp [hunk 3]
```diff
@@ -453,7 +460,12 @@ struct d3d12_buffer {
         ReadBack
     } type;
 
-    halide_buffer_t *halide;
+    // UNSAFE to keep a pointer to a 'halide_buffer_t' since it is a POD type
+    // that can be re-assigned / re-scoped outside of this runtime module...
+    //halide_buffer_t *halide;
```
Comment (shoaibkamil) [inline src/runtime/d3d12compute.cpp]: Remove this line?

File: src/runtime/hashmap.h (hunk 1 above)
Comment (slomp) [reply src/runtime/hashmap.h]: Yes. This way, whichever runtime module that needs to include `hashmap.h` has a chance to override the default malloc/free implementation.
That's what `d3d12compute.cpp` is doing:
https://github.com/halide/Halide/blob/dbf75159454a6380a41eef7e2ee90922919ee423/src/runtime/d3d12compute.cpp#L200-L202

File: src/runtime/d3d12compute.cpp [hunk 4]
```diff
@@ -453,7 +460,12 @@ struct d3d12_buffer {
...
+    // keeping a copy of it is a bit overkill, so we'll use a flag:
+    bool halide;
```
Comment (shoaibkamil) [inline src/runtime/d3d12compute.cpp]: I think there could be a more descriptive name than "halide".

File: src/runtime/d3d12compute.cpp (hunk 2 above)
Comment (slomp) [reply src/runtime/d3d12compute.cpp]: 👍

File: src/runtime/d3d12compute.cpp [hunk 5]
```diff
@@ -667,14 +679,28 @@ void release_d3d12_object<d3d12_compute_pipeline_state>(d3d12_compute_pipeline_s
...
+
 WEAK int wrap_buffer(struct halide_buffer_t *hbuffer, d3d12_buffer *dbuffer) {
     halide_assert(user_context, (hbuffer->device == 0));
     if (hbuffer->device != 0) {
         return -2;
     }
 
     halide_assert(user_context, (dbuffer->resource != NULL));
-    halide_assert(user_context, (dbuffer->halide == NULL));
+    halide_assert(user_context, (!dbuffer->halide));
```
Comment (shoaibkamil) [inline src/runtime/d3d12compute.cpp]: I think these two lines should be consistently styled; I prefer the more explicity `== NULL` but not sure what is appropriate Halide coding style.

File: src/runtime/d3d12compute.cpp [hunk 6]
```diff
@@ -1897,6 +1839,31 @@ static void *buffer_contents(d3d12_buffer *buffer) {
             pData = map_buffer(&readback);
             break;
 
+        /*
+        // TODO(marcos): this is probably how these buffer types should be handled:
```
Comment (shoaibkamil) [inline src/runtime/d3d12compute.cpp]: Why aren't they handled this way?

File: src/runtime/d3d12compute.cpp [hunk 7]
```diff
@@ -2246,7 +2343,7 @@ WEAK int halide_d3d12compute_device_sync(void *user_context, struct halide_buffe
 
     #ifdef DEBUG_RUNTIME
     uint64_t t_after = halide_current_time_ns(user_context);
-    debug(user_context) << TRACEINDENT << "Time for halide_d3d12compute_device_sync: " << (t_after - t_before) / 1.0e6 << " ms\n";
+    TRACEPRINT( "Time for halide_d3d12compute_device_sync: " << (t_after - t_before) / 1.0e6 << " ms\n" );
```
Comment (shoaibkamil) [inline src/runtime/d3d12compute.cpp]: Extraneous space after/before parens.

File: src/runtime/d3d12compute.cpp (hunk 4 above)
Comment (steven-johnson) [reply src/runtime/d3d12compute.cpp]: definitely needs a better name. I have no idea what this means.

File: src/runtime/d3d12compute.cpp (hunk 6 above)
Comment (steven-johnson) [reply src/runtime/d3d12compute.cpp]: yeah, this needs better explanation about why not

File: src/runtime/hashmap.h [hunk 8]
```diff
@@ -6,6 +6,14 @@
 #include "printer.h"
 #include "scoped_mutex_lock.h"
 
+#ifndef hashmap_malloc
+#define hashmap_malloc(size)    halide_malloc(NULL, size)
+#endif//hashmap_malloc
```
Comment (steven-johnson) [inline src/runtime/hashmap.h]: Two spaces between #endif and //

File: src/runtime/hashmap.h [hunk 9]
```diff
@@ -6,6 +6,14 @@
...
+
+#ifndef hashmap_free
+#define hashmap_free(memory)    halide_free(NULL, memory)
+#endif//hashmap_free
```
Comment (steven-johnson) [inline src/runtime/hashmap.h]: Two spaces between #endif and //

File: src/runtime/hashmap.h (hunk 8 above)
Comment (steven-johnson) [reply src/runtime/hashmap.h]: halide_malloc and halide_free take a user_context; clients that override these may fail if you pass in null. You need to add user_context as arguments to the hashmap_malloc/free macros and pass them everywhere. (If that isn't possible, you can't use halide_malloc.)

File: src/runtime/d3d12compute.cpp [hunk 10]
```diff
@@ -2071,9 +2038,7 @@ extern "C" {
 WEAK int halide_d3d12compute_device_malloc(void *user_context, halide_buffer_t *buf) {
     TRACELOG;
 
-    debug(user_context) << TRACEINDENT
-                        << "(user_context: " << user_context
-                        << ", buf: " << buf << ")\n";
+    TRACEPRINT( "(user_context: " << user_context << ", buf: " << buf << ")\n" );
```
Comment (shoaibkamil) [inline src/runtime/d3d12compute.cpp]: Extraneous spaces

File: src/runtime/d3d12compute.cpp [hunk 11]
```diff
@@ -2087,7 +2052,7 @@ WEAK int halide_d3d12compute_device_malloc(void *user_context, halide_buffer_t *
         halide_assert(user_context, buf->dim[i].stride > 0);
     }
 
-    debug(user_context) << TRACEINDENT << "allocating " << *buf << "\n";
+    TRACEPRINT( "allocating " << *buf << "\n" );
```
Comment (shoaibkamil) [inline src/runtime/d3d12compute.cpp]: Extraneous spaces

File: src/runtime/d3d12compute.cpp [hunk 12]
```diff
@@ -2111,7 +2076,7 @@ WEAK int halide_d3d12compute_device_malloc(void *user_context, halide_buffer_t *
 
     #ifdef DEBUG_RUNTIME
     uint64_t t_after = halide_current_time_ns(user_context);
-    debug(user_context) << TRACEINDENT << "Time: " << (t_after - t_before) / 1.0e6 << " ms\n";
+    TRACEPRINT( "Time: " << (t_after - t_before) / 1.0e6 << " ms\n" );
```
Comment (shoaibkamil) [inline src/runtime/d3d12compute.cpp]: Extraneous spaces

File: src/runtime/d3d12compute.cpp [hunk 13]
```diff
@@ -2136,7 +2102,7 @@ WEAK int halide_d3d12compute_device_free(void *user_context, halide_buffer_t *bu
...
```
Comment (shoaibkamil) [inline src/runtime/d3d12compute.cpp]: Extraneous spaces

File: src/runtime/d3d12compute.cpp [hunk 14]
```diff
@@ -2179,21 +2145,148 @@ WEAK int halide_d3d12compute_initialize_kernels(void *user_context, void **state
 
         #ifdef DEBUG_RUNTIME
         uint64_t t_after_compile = halide_current_time_ns(user_context);
-        debug(user_context) << TRACEINDENT << "Time for halide_d3d12compute_initialize_kernels compilation: " << (t_after_compile - t_before_compile) / 1.0e6 << " ms\n";
+        TRACEPRINT( "Time for halide_d3d12compute_initialize_kernels compilation: " << (t_after_compile - t_before_compile) / 1.0e6 << " ms\n" );
```
Comment (shoaibkamil) [inline src/runtime/d3d12compute.cpp]: Extraneous spaces... basically change this throughout :)

File: src/runtime/d3d12compute.cpp (hunk 4 above)
Comment (slomp) [reply src/runtime/d3d12compute.cpp]: Yeap, that's a left-over; `halide` used to be a `halide_buffer_t*` cached within a `d3d12_buffer` for convenience. After I realized it was unsafe to cache such a pointer, I replaced it for a `bool` to minimize the number o changes in the code at that time, but in reality, this field is not necessary at all. It's gone now.

File: src/runtime/d3d12compute.cpp (hunk 6 above)
Comment (slomp) [reply src/runtime/d3d12compute.cpp]: I enabled those cases, but they will never get called in the current back-end implementation. I protected with a (paranoid) assert, just in case.

File: src/runtime/hashmap.h (hunk 1 above)
Comment (slomp) [reply src/runtime/hashmap.h]: I added some comments above it explaining the rationale of it.

File: src/runtime/d3d12compute.cpp [hunk 15]
```diff
@@ -82,13 +81,18 @@ class StackPrinter : public Printer<type, length> {
 static void *const user_context = NULL;
 //
 #if HALIDE_D3D12_TRACE
-    #ifndef DEBUG_RUNTIME
-    #define debug(x) Printer<BasicPrinter>(x)
-    #endif//DEBUG_RUNTIME
```
[the remark this answers is not in the document]
Comment (slomp) [reply src/runtime/d3d12compute.cpp]: That code has been removed already :)

File: src/runtime/hashmap.h (hunk 8 above)
Comment (slomp) [reply src/runtime/hashmap.h]: I am now preserving the `user_context` around it.

However, I should point that `cache.cpp` (which is from I based the `hashmap.h` code), `halide_malloc()` and `halide_free()` are used without `user_context`.
(I think `cache.cpp` is @zvookin's domain)

File: src/Target.cpp [hunk 16]
```diff
@@ -16,6 +16,10 @@
 #include <sys/auxv.h>
 #endif
 
+#ifdef  _MSC_VER
+#include <intrin.h>
+#endif  //_MSC_VER
```
Comment (steven-johnson) [inline src/Target.cpp]: Please put a space after // here and everywhere else

File: src/runtime/d3d12compute.cpp [hunk 17]
```diff
@@ -667,14 +678,26 @@ void release_d3d12_object<d3d12_compute_pipeline_state>(d3d12_compute_pipeline_s
...
+}
+
+static const d3d12_buffer *peel_buffer(const struct halide_buffer_t *hbuffer) {
+    return peel_buffer(const_cast<halide_buffer_t*>(hbuffer));
+}
+
 WEAK int wrap_buffer(struct halide_buffer_t *hbuffer, d3d12_buffer *dbuffer) {
     halide_assert(user_context, (hbuffer->device == 0));
     if (hbuffer->device != 0) {
         return -2;
```
Comment (steven-johnson) [inline src/runtime/d3d12compute.cpp]: Should this be a halide_error_code_ instead?

File: src/runtime/d3d12compute.cpp [hunk 18]
```diff
@@ -1897,8 +2076,32 @@ static void *buffer_contents(d3d12_buffer *buffer) {
             pData = map_buffer(&readback);
             break;
 
-        case d3d12_buffer::Unknown :
+        case d3d12_buffer::ReadOnly  :
```
Comment (steven-johnson) [inline src/runtime/d3d12compute.cpp]: aligning the colons is weird and I don't think we do it elsewhere. Just make them flush.

**Comment by slomp** [comment]
All known issues have been fixed.
It's ready to merge after the review.
We'll need to re-enable d3d12 in the automated tests too.

File: src/runtime/d3d12compute.cpp [hunk 19]
```diff
@@ -2608,48 +2790,189 @@ WEAK int halide_d3d12compute_run(void *user_context,
...
+    // allocating the device memory and MapBuffering it to the host memory, but
+    // sadly, even with a "dedicated" d3d12 staging heap just for this buffer,
+    // d3d12 has no bi-directional system memory heap (a sigle resource that is
+    // capable of both upload and readback). One possible workaround is to just
+    // suballocate from two common heaps (one used for uploads and another for
+    // readbacks) and dynamically change buffer->host accordingly. However, if
+    // the user ever caches the buffer->host pointer, that would be really bad!
+    //
+    // Another complicating factor is the fact that the test "cleanup_on_error"
+    // expects hallide_malloc() to be called here to allocate the host buffer.
```
Comment (steven-johnson) [inline src/runtime/d3d12compute.cpp]: s/hallide/halide/

**Comment by shoaibkamil** [comment]
All my comments have been addressed, but please sync with master; Steven's fix may have fixed the FFT crashes.

File: src/runtime/d3d12compute.cpp [hunk 20]
```diff
@@ -1870,23 +1806,266 @@ static void wait_until_completed(d3d12_compute_command_list *cmdList) {
...
+    halide_assert(user_context, num_bytes > 0);
+
+    if (src->type == d3d12_buffer::Upload) {
+        // host-to-device via staging buffer:
+        halide_assert(user_context, (dst->type != d3d12_buffer::Upload));
+        halide_assert(user_context, (dst->type != d3d12_buffer::ReadBack));
+        halide_assert(user_context, (dst->xfer == NULL));
+        // TODO: assert that offsets and sizes are within bounds
+        d3d12_buffer::transfer_t xfer = { };
+            xfer.staging = src;
```
Comment (steven-johnson) [inline src/runtime/d3d12compute.cpp]: Why the weird indentation?

File: src/runtime/d3d12compute.cpp [hunk 21]
```diff
@@ -1870,23 +1806,266 @@ static void wait_until_completed(d3d12_compute_command_list *cmdList) {
...
+    }
+
+    if (dst->type == d3d12_buffer::ReadBack) {
+        // device-to-host via staging buffer:
+        halide_assert(user_context, (src->type != d3d12_buffer::Upload));
+        halide_assert(user_context, (src->type != d3d12_buffer::ReadBack));
+        halide_assert(user_context, (dst->xfer == NULL));
+        // TODO: assert that offsets and sizes are within bounds
+        d3d12_buffer::transfer_t xfer = { };
+            xfer.staging = dst;
```
Comment (steven-johnson) [inline src/runtime/d3d12compute.cpp]: Why the weird indentation?

**Comment by shoaibkamil** [comment]
This includes the removed FFT stack canary, so I'm not sure why FFT is still crashing.  I can't replicate on my local machine either.
